luigi/batch_notifier.py

```diff
@@ -0,0 +1,201 @@
+    batch_mode = luigi.parameter.ChoiceParameter(
+        default='unbatched_params', choices=('family', 'all', 'unbatched_params'),
+        description='Method used for batching failures in e-mail. If "family" all failures for '
+                    'tasks with the same family will be batched. If "unbatched_params", all '
+                    'failures for tasks with the same family and non-batched parameters will be '
+                    'batched. If "all", tasks will only be batched if they have identical names.')
+    error_lines = luigi.parameter.IntParameter(
+        default=20, description='Number of lines to show from each error message. 0 means show all')
+    error_messages = luigi.parameter.IntParameter(
+        default=1, description='Number of error messages to show for each group')
+    group_by_error_messages = luigi.parameter.BoolParameter(
+        default=True, description='Group items with the same error messages together')
+
+
+class ExplQueue(collections.OrderedDict):
+    def __init__(self, num_items):
+        self.num_items = num_items
+        super(ExplQueue, self).__init__()
+
+    def enqueue(self, item):
+        self.pop(item, None)
+        self[item] = datetime.now()
+        if len(self) > self.num_items:
+            self.popitem(last=False)  # pop first item if past length
+
+
+def _fail_queue(num_messages):
+    return lambda: collections.defaultdict(lambda: ExplQueue(num_messages))
+
+
+def _plural_format(template, number, plural='s'):
+    if number == 0:
+        return ''
+    return template.format(number, '' if number == 1 else plural)
+
+
+class BatchNotifier(object):
+    def __init__(self, **kwargs):
+        self._config = batch_email(**kwargs)
+        self._fail_counts = collections.defaultdict(collections.Counter)
+        self._disabled_counts = collections.defaultdict(collections.Counter)
+        self._scheduling_fail_counts = collections.defaultdict(collections.Counter)
+        self._fail_expls = collections.defaultdict(_fail_queue(self._config.error_messages))
+        self._update_next_send()
+
+        self._email_format = email().format
+        if email().receiver:
+            self._default_owner = set(filter(None, email().receiver.split(',')))
+        else:
+            self._default_owner = set()
+
+    def _update_next_send(self):
+        self._next_send = time.time() + 60 * self._config.email_interval
+
+    def _key(self, task_name, family, unbatched_args):
+        if self._config.batch_mode == 'all':
+            return task_name
+        elif self._config.batch_mode == 'family':
+            return family
+        elif self._config.batch_mode == 'unbatched_params':
+            param_str = ', '.join('{}={}'.format(*kv) for kv in six.iteritems(unbatched_args))
+            return '{}({})'.format(family, param_str)
+        else:
+            raise ValueError('Unknown batch mode for batch notifier: {}'.format(
+                self._config.batch_mode))
```

Comment:
This could happen if you add a new choice but forget to implement it. It should never happen with the code as-is, but things change.